torch/package/find_file_dependencies.py

```diff
@@ -57,47 +61,53 @@ def _grab_node_str(self, node):
     def visit_Call(self, node):
         # __import__ calls aren't routed to the visit_Import/From nodes
         if hasattr(node.func, "id") and node.func.id == "__import__":
-            if type(node.args[0]) not in [ast.Constant, ast.Str]:
-                # We don't want to parse dynamic uses of __import__
+            try:
+                name = self._grab_node_str(node.args[0])
+                fromlist = []
+                level = 0
+                if len(node.args) > 3:
+                    for v in node.args[3].elts:
+                        fromlist.append(self._grab_node_str(v))
+                elif hasattr(node, "keywords"):
+                    for keyword in node.keywords:
+                        if keyword.arg == "fromlist":
+                            for v in keyword.value.elts:
+                                fromlist.append(self._grab_node_str(v))
+                if len(node.args) > 4:
+                    level = self._grab_node_int(node.args[4])
+                elif hasattr(node, "keywords"):
+                    for keyword in node.keywords:
+                        if keyword.arg == "level":
+                            level = self._grab_node_int(keyword.value)
+                if fromlist == []:
+                    # the top-level package (the name up till the first dot) is returned
+                    # when the fromlist argument is empty in normal import system,
+                    # we need to include top level package to match this behavior and last
+                    # level package to capture the intended dependency of user
+                    self.references[(name, None)] = True
+                    top_name = name.rsplit(".", maxsplit=1)[0]
+                    if top_name != name:
+                        top_name = self._absmodule(top_name, level)
+                        self.references[(top_name, None)] = True
+                else:
+                    name = self._absmodule(name, level)
+                    for alias in fromlist:
+                        # fromlist args may be submodules, so we have to add the fromlist args
+                        # to the list of potential references. If import of an arg fails we
+                        # will ignore it, similar to visit_ImportFrom
+                        if alias != "*":
+                            self.references[(name, alias)] = True
+                        else:
+                            self.references[(name, None)] = True
+            except Exception as e:
+                info_string = (
+                    "module: '"
+                    + self.module
+                    + "', lineno: "
+                    + str(node.__dict__["lineno"])
+                )
+                self.failed_dunder_imports.append(info_string)
```

Comment:
Why not put this in the `else` branch corresponding to `if type(node.args[0]) not in [ast.Constant, ast.Str]`? Do we want to try parsing dynamic uses of `__import__` in case some can be successfully parsed?

Reply:
This AST parsing code is very brittle. I wouldn't be surprised if there are more errors than just ones related to dynamic imports. I think being conservative here by catching all errors will lead to a better UX